### PR TITLE
Small improvements for DFConv and Axi4 related stuff

### PR DESCRIPTION
**Pull request by samalws**
This PR adds a bunch of small improvements for DfConv and Axi4 related stuff:
- Fixes the AXI4 DfConv instances so that the Write instance doesn't infinite loop and the Read instance doesn't duplicate data
- Adds tests for the AXI4 instances
- Moves the tests around into more sensible filenames
- Renames `GoodAxi4...Config`s to `KnownAxi4...Config`
- Adds some instance derivations to Axi4 files needed to make the tests typecheck
- Adds an `instance DfConv (Reverse a)` and a test to make sure it works

### CODE REVIEW

File: tests/Tests/Protocols/Axi4.hs [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+        NonBufferable
+        NonModifiable
+        OtherNoLookupCache
+        NoLookupCache ))
+    (toKeepType  (T3
+        Privileged
+        Secure
+        Instruction ))
+    (toKeepType 0)
+    0
```
Comment (martijnbastiaan) [inline tests/Tests/Protocols/Axi4.hs:250]: We should use record names here

File: tests/Tests/Protocols/Axi4.hs [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+ where
+  ckt :: (C.HiddenClockResetEnable dom) =>
+   Circuit
+   (Axi4ReadAddress dom ConfAR Int,
+    Reverse (Axi4ReadData dom ConfR Int Int))
+   (Axi4ReadAddress dom ConfAR Int,
+    Reverse (Axi4ReadData dom ConfR Int Int))
+  ckt = DfConv.convert Proxy Proxy
+
+  genInfo = (,,) <$> DfTest.genSmallInt <*> DfTest.genSmallInt <*> (toKeepType <$> (pure ROkay C.<|> pure RExclusiveOkay C.<|> pure RSlaveError C.<|> pure RDecodeError))
```
Comment (martijnbastiaan) [inline tests/Tests/Protocols/Axi4.hs:229]: Split this up over multiple lines

**Comment by samalws** [comment]
Ok done

**Comment by martijnbastiaan** [comment]
Thanks!

**Pull request merged**